### PR TITLE
Satellite-Graphs / Broken Sorted Gather

### PR DESCRIPTION
**Pull request by maierlars**
Added new optmizer rule to fix broken sorted gather when used together with satellite collections.

### CODE REVIEW

File: arangod/Aql/OptimizerRules.cpp [hunk 1]
```diff
@@ -7686,3 +7677,31 @@ void arangodb::aql::spliceSubqueriesRule(Optimizer* opt, std::unique_ptr<Executi
 
   opt->addPlan(std::move(plan), rule, modified);
 }
+
+void arangodb::aql::decayUnnecessarySortedGather(Optimizer* opt, std::unique_ptr<ExecutionPlan> plan,
```
Comment (jsteemann) [inline arangod/Aql/OptimizerRules.cpp]: we normally indent with just 2 spaces. why is it all different here?

File: js/common/modules/@arangodb/aql/explainer.js [hunk 2]
```diff
@@ -1702,7 +1702,10 @@ function processQuery(query, explain, planIndex) {
         }
         if (node.sortmode !== 'unset') {
           gatherAnnotations.push('sort mode: ' + node.sortmode);
-        } 
+        }
+        if (node.elements.length === 0) {
+          gatherAnnotations.push('unsorted');
+        }
```
Comment (jsteemann) [inline js/common/modules/@arangodb/aql/explainer.js]: Is it possible that the sort mode is set to "sorted" but that there are 0 elements, so we will end up with `sort mode: sorted, unsorted` later?

Comment (goedderz) [reply js/common/modules/@arangodb/aql/explainer.js]: No, the decision whether `sortmode` will be set is exactly based on whether `elements` is empty or not. So this should be correct.

Comment (jsteemann) [reply js/common/modules/@arangodb/aql/explainer.js]: Ok

File: arangod/Aql/ClusterNodes.cpp [hunk 3]
```diff
@@ -423,20 +423,28 @@ CostEstimate DistributeNode::estimateCost() const {
...
+          return enumNode->collection();
+        }
+        break;
+      }
       case INDEX:
         return castTo<IndexNode const*>(node)->collection();
       case TRAVERSAL:
       case SHORTEST_PATH:
       case K_SHORTEST_PATHS:
         return castTo<GraphNode const*>(node)->collection();
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp]: Don't we need the same change (skipping satellites) not just for `ENUMERATE_COLLECTION`, but also for `INDEX` and the GraphNodes?
Also, you can combine the `ENUMERATE_COLLECTION` and `INDEX` cases by casting them to `CollectionAccessingNode`s.

File: tests/js/server/aql/aql-optimizer-rule-decay-unnecessary-sorted-gather.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+const db = require("@arangodb").db;
+const jsunity = require("jsunity");
+const console = require('console');
+const _ = require("lodash");
+const isEnterprise = require("internal").isEnterprise();
+
+function optimizerRuleTestSuite() {
+  const ruleName = "decay-unnecessary-sorted-gather";
+
+  const sharedCollection = "c";
```
Comment (goedderz) [inline tests/js/server/aql/aql-optimizer-rule-decay-unnecessary-sorted-gather.js]: Is that intentionally _shared_, or did you mean to write _sharded_?

File: tests/js/server/aql/aql-optimizer-rule-decay-unnecessary-sorted-gather.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+      db._drop(sharedCollection);
+      db._drop(singleShardCollection);
+      db._drop(satelliteCollection);
+      db._drop(smartJoinCollection2);
+      db._drop(smartJoinCollection1);
+    },
+
+    testNonEnterprise: function () {
+      // this query has to be that complicated, because
+      // 1. if we only use a single collection with repl fact 1, the one shard rule will apply
```
Comment (goedderz) [inline tests/js/server/aql/aql-optimizer-rule-decay-unnecessary-sorted-gather.js]: You could just disable the `cluster-one-shard` rule.

File: tests/js/server/aql/aql-optimizer-rule-decay-unnecessary-sorted-gather.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+      // 2. the FILTER forces the sharded collection to enumerated first
```
Comment (goedderz) [inline tests/js/server/aql/aql-optimizer-rule-decay-unnecessary-sorted-gather.js]: You could disable the `interchange-adjacent-enumeration` rule, which I suppose switched the enumerations.

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9188/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9225/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9229/

File: arangod/Aql/ClusterNodes.cpp [hunk 7]
```diff
@@ -423,20 +423,26 @@ CostEstimate DistributeNode::estimateCost() const {
...
+        auto const* cNode = castTo<CollectionAccessingNode const*>(node);
+        if (!cNode->isUsedAsSatellite() && cNode->prototypeCollection() == nullptr) {
+          return cNode->collection();
+        }
+        break;
+      }
       case TRAVERSAL:
       case SHORTEST_PATH:
       case K_SHORTEST_PATHS:
         return castTo<GraphNode const*>(node)->collection();
```
Comment (goedderz) [inline arangod/Aql/ClusterNodes.cpp:437]: @maierlars 
I still think this is not correct for graphs. First, this must always be a local graph node (i.e. on a DBServer). Thus it is also a `CollectionAccessingNode`. You have to cast it and take the collection then, but only if it is not used as a satellite.

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9416/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9462/

**Comment by mchacki** [comment]
Jenkins blue besides known issue:
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/9502/

**Pull request merged**